### ISSUE TITLE
Cannot configure smtp for outgoing email through Office365

### ISSUE THREAD

**Issue by nicon8** [metabase/metabase#4272]
Hi all,

I'm try to configure smtp with office 365 from a docker container. 

I'm using the currently latest version 0.22.2 on docker 1.13.0 running on ubuntu 14.04. If I set any smtp security I cannot save the settings and I obtain the error "Sorry, something went wrong. Please try again."
The mail server is reachable by the container. 

The container logs are:

01-31 07:50:33 ERROR metabase.email :: Error testing SMTP connection: 
01-31 07:50:33 WARN api.email :: Problem connecting to mail server: 
01-31 07:50:33 ERROR metabase.middleware :: PUT /api/email 500 (2 s) (0 DB calls)
{:message "Sorry, something went wrong.  Please try again."}

Is there a way to delete mail setting after save? I try without security which is wrong but now I cannot add new users because the system wants to invite them by mail (which do not work)

Thanks,
Nicola

**Comment by tlrobinson** [comment]
@nicon8 Unfortunately it's difficult for us to debug deployment environment issues like this. I haven't had any issues with SMTP in Docker. Could you try just running the jar directly and seeing if you're able to get it to work there, or try a different SMTP server on your Docker instance?

**Comment by nicon8** [comment]
Hi, thanks for your answer. 
The problem seems related to office365 smtp server. The google smtp server works. 

Office365 does not work even running the jar directly, so we can conclude is not definitively an issue related to docker, but to office365 smtp server. 

The google smtp server works running the jar directly and works inside a docker container. Sorry for the wrong issue summary.

**Comment by temal-** [comment]
Hi,

I ran into the same problem and found a way to reproduce the issue with and without valid credentials:

- `docker run -ti -p 3000:3000 --name metabase metabase/metabase` (currently v0.22.2, Built on 2017-01-26)
- Create an account, skip everything and open eMail in the admin settings
- Enter `smtp.office365.com` as SMTP HOST and `587` as SMTP PORT

There are two different choices here:

## SSL

If you choose SSL it doesn't matter what you put into the username, password and from-address field since you get the following message on SMTP HOST and SMTP PORT immediately after a click on Save: `Wrong host or port`

Logfile:

```
03-14 11:54:04 ERROR metabase.email :: Error testing SMTP connection: Could not connect to SMTP host: smtp.office365.com, port: 587
03-14 11:54:04 WARN api.email :: Problem connecting to mail server: Could not connect to SMTP host: smtp.office365.com, port: 587
03-14 11:54:04 ERROR metabase.middleware :: PUT /api/email 500 (167 ms) (0 DB calls)
{:errors {:email-smtp-host "Wrong host or port", :email-smtp-port "Wrong host or port"}}

```

## TLS

If you choose TLS it takes a moment longer but then ends with: `Sorry, something went wrong. Please try again.`.

Logfile:

```
03-14 11:59:44 ERROR metabase.email :: Error testing SMTP connection:
03-14 11:59:44 WARN api.email :: Problem connecting to mail server:
03-14 11:59:44 ERROR metabase.middleware :: PUT /api/email 500 (2 s) (0 DB calls)
{:message "Sorry, something went wrong.  Please try again."}
```

I tested SSL and TLS with valid and also with invalid credentials and discovered the exact same behaviour.

I'm not sure what the TLS error is but there is something wrong with SSL since denying the host+port combination of office365 shouldn't be the way to convince people to use TLS. ;)

Thanks in advance,
Malte

**Comment by pablohere** [comment]
Seeing the exact same issue on my end.

**Comment by salsakran** [comment]
Is anyone hitting this issue on an smtp server *other* than outlook?

We don't use outlook, so we'll need some help from the community on this one.

**Comment by stepan-romankov** [comment]
Hi, the same problem.
Connection via **telnet** from inside container to **smtp.office365.com 587** is fine.

**Comment by optiomas** [comment]
Also seeing this.

**Comment by thearthur** [comment]
email authentication has four modes with non-intuitive names:

1. smtp aka; plaintext, none, unauthenticated, the-old-one, and many other names. This often sends the password in the clear across the internet, or doesn't use a password at all. 
2. SSL (aka the "new old one") this is authenticated and uses a username+password or certificate
3. TLS this is the "new version" of SSL. it has more security-sauce and you should use it where it works
4. STARTTLS this is a method of doing smtp like #1 without sending the password or message in the clear. This protocol usually shares a port with smtp so it's totally unclear when this option is required. Many servers, including it seems Office365, use this option for securing email connections.

the short version is that I need to add these options to the email send function
```
"mail.smtp.starttls.enable" "true"
"mail.smtp.starttls.required" "true"
```
more details soon.

**Comment by thearthur** [comment]
the add-starttls-smtp-authentication branch ads a new authentication type called "STARTTLS" that sets the appropriate smtp options to use the starttls protocol. Next I'm looking into adding an option to detect when this is required:

There are a couple of options:
1. Offer four options and not try to be smart about it.
2. add a fifth option for "AUTO" and try the options in order from most to least secure till one works, then save this automatically. This would not consider smtp/insecure. If none of them work, then it will display the "there was a problem" message.
3. If the connection fails, try the other security options and silently correct the problem If possible.
4. Offer only two security settings "AUTO" and "INSECURE". This may cause problems for cases where we think the connection worked but it's not able to send email. For instance if a site offers TLS and SSL, and the TLS server is badly configured. People need a way to correct it.
5. Offer only automatic detection and don't allow insecure connections. This is included here only for completeness, it would break for people using insecure internal mail servers.
6. Always auto-detect, and silently downgrade to smtp/insecure. This could downgrade people's security if they enter a wrong password or a server is momentarily down. It's a bad idea.

**Comment by salsakran** [comment]
How about 3 + a error-looking notification on save? 

I.e. if the user enters some creds, we try to use them faithfully. If there is a minor modification (only to other encrypted protocols or ports), we tell the user, and save it in the working config.

The rest seem a bit too opaque.

**Comment by NewEvolution** [comment]
Still seeing the identical issue when attempting to send via Office365

```
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:170)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
	at sun.security.ssl.InputRecord.read(InputRecord.java:503)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at com.sun.mail.util.SocketFetcher.configureSSLSocket(SocketFetcher.java:598)
	at com.sun.mail.util.SocketFetcher.createSocket(SocketFetcher.java:372)
	at com.sun.mail.util.SocketFetcher.getSocket(SocketFetcher.java:238)
	at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2066)
	... 100 more
07-12 15:52:14 WARN api.email :: Problem connecting to mail server: 
07-12 15:52:14 ERROR metabase.middleware :: PUT /api/email 500 (20 s) (0 DB calls).
{:message "Sorry, something went wrong.  Please try again."}
```

Is the recommended fix really "don't use Office365" as per https://www.metabase.com/troubleshooting/email/ ?

**Comment by davidbonachera** [comment]
So for the one struggling, if you selected STARTTLS first and you are trying to change it to TLS you'll always get the error afterward.

The way I solved the problem.
1. Click on clear to clean the data.
2. Setup the SMTP with smtp.office365 and use TLS.
